stdlib/LinearAlgebra/src/adjtrans.jl

```diff
@@ -378,15 +378,31 @@ Broadcast.broadcast_preserving_zero_d(f, tvs::Union{Number,TransposeAbsVec}...)
 
 
 ### reductions
-# faster to sum the Array than to work through the wrapper
-Base._mapreduce_dim(f, op, init::Base._InitialValue, A::Transpose, dims::Colon) =
+# faster to sum the Array than to work through the wrapper (but only in commutative reduction ops)
+const CommutativeOps = Union{typeof(+),typeof(Base.add_sum),typeof(min),typeof(max),typeof(|),typeof(&)}
+Base._mapreduce_dim(f, op::CommutativeOps, init::Base._InitialValue, A::Transpose, dims::Colon) =
     transpose(Base._mapreduce_dim(_sandwich(transpose, f), _sandwich(transpose, op), init, parent(A), dims))
-Base._mapreduce_dim(f, op, init::Base._InitialValue, A::Adjoint, dims::Colon) =
+Base._mapreduce_dim(f, op::CommutativeOps, init::Base._InitialValue, A::Adjoint, dims::Colon) =
     adjoint(Base._mapreduce_dim(_sandwich(adjoint, f), _sandwich(adjoint, op), init, parent(A), dims))
```

Comment:
It seems more intuitive to make these `CommutativeOps` call
```julia
Base._mapreduce_dim(f, op::CommutativeOps, init::Base._InitialValue, A::Transpose, dims::Colon) =
    Base._mapreduce_dim(f∘transpose, op, init, parent(A), dims)
```
Then we can add `Base._extrema_rf` to this list to fix #46714

Reply:
And L399 could be replaced with
```julia
Base.mapreducedim!(f, op::CommutativeOps, B::AbstractArray, A::TransposeAbsMat) =
    (Base.mapreducedim!(f∘transpose, op, B, parent(A)); B)
Base.mapreducedim!(f, op::CommutativeOps, B::AbstractArray, A::AdjointAbsMat) =
    (Base.mapreducedim!(f∘transpose, op, B, parent(A)); B)
switch_dim12(B::AbstractVector) = permutedims(B)
switch_dim12(B::AbstractArray{<:Any,0}) = B
switch_dim12(B::AbstractArray) = PermutedDimsArray(B, (2, 1, ntuple(Base.Fix1(+,2), ndims(B) - 2)...))
```

Reply:
> ```julia
> Base._mapreduce_dim(f, op::CommutativeOps, init::Base._InitialValue, A::Transpose, dims::Colon) =
>     Base._mapreduce_dim(f∘transpose, op, init, parent(A), dims)
> Base._mapreduce_dim(f, op::CommutativeOps, init::Base._InitialValue, A::Adjoint, dims::Colon) =
>     Base._mapreduce_dim(f∘adjoint, op, init, parent(A), dims)
> ```

The sandwiching mechanism was introduced by @mcabbott in #39513, and I believe it is for the following reason: summing a transpose matrix of (recursively transposed) matrices should perhaps result in a `Transpose`, which is what the current design yields. Otherwise redirecting to `f∘transpose` - perhaps in more restricted (`Number`?) cases - should be good enough.

Reply:
Oh that's embarrassing, sorry, I thought we thought this all through carefully. But `prod` is clearly broken.

Reply:
OK, now I think the sandwich story is just wrong, even for arrays of numbers:
```julia
julia> mapreduce(imag, +, [1 2im 3])
2

julia> mapreduce(imag, +, [1, 2im, 3]')
-2

julia> mapreduce(x -> [x;;;], +, [1 2 3])
1×1×1 Array{Int64, 3}:
[:, :, 1] =
 6

julia> mapreduce(x -> [x;;;], +, [1, 2, 3]')
ERROR: adjoint not defined for Array{Int64, 3}. 
```

The suggestion by N5N3 to do `f∘transpose` seems like a good one to me. 

Is the restriction to `CommutativeOps` necessary? `reduce` makes no claims about the order of operations, so arguably `prod(::Matrix{Matrix})` is ill-defined, you should call `foldl`. The case of  `prod(::Adjoint{Float64, ...})` is one that it would be nice to send the fast path.

Reply:
Since these dispatches are for performance reason, I think we'd better keep the same result as the general fallback.
As for `prod(::Adjoint{Float64, ...})`, `*` should be commutative for `Number`, maybe we can use `Core.Compiler,return_type` to turn the optimization on.
It should be ok as even the inference fails, the result should be similar (with negligible numerical difference).

Reply:
> Is the restriction to `CommutativeOps` necessary? `reduce` makes no claims about the order of operations, 

For, say, `reduce(*, ["a", "b", "c", "d"])` all of the following would be valid ways to compute the result:
* `"a" * ("b" * ("c" * "d"))`
* `(("a" * "b") * "c") * "d"`
* `("a" * "b") * ("c" * "d")`
* `("a" * ("b" * "c")) * "d"`
* `"a" * (("b" * "c") * "d")`

They all yield `"abcd"`. But `reduce` is not allow to do, say `"a" * (("c" * "b") * "d")`. But if the operation is commutative, the result is the same, so `reduce` would be ok to do this reordering then.